libsignal-protocol/src/crypto.rs

```diff
@@ -56,49 +299,185 @@ impl CryptoProvider {
+
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(key, key_len);
+    user_data.0.hmac_sha256_init(buffer).into_code()
 }
 
 unsafe extern "C" fn hmac_sha256_update_func(
     _hmac_context: *mut c_void,
-    _data: *const u8,
-    _data_len: usize,
-    _user_data: *mut c_void,
+    data: *const u8,
+    data_len: usize,
+    user_data: *mut c_void,
 ) -> i32 {
-    unimplemented!()
+    assert!(!data.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(data, data_len);
+    user_data.0.hmac_sha256_update(buffer).into_code()
 }
 
-unsafe extern "C" fn random_func(
-    data: *mut u8,
-    len: usize,
+unsafe extern "C" fn sha512_digest_init_func(
+    _digest_context: *mut *mut c_void,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    user_data.0.sha512_digest_init().into_code()
+}
+
+unsafe extern "C" fn sha512_digest_update_func(
+    _digest_context: *mut c_void,
+    data: *const u8,
+    data_len: usize,
     user_data: *mut c_void,
 ) -> c_int {
     assert!(!data.is_null());
     assert!(!user_data.is_null());
 
-    let user_data = &*(user_data as *const State);
-    let buffer = slice::from_raw_parts_mut(data, len);
-    user_data.0.fill_random(buffer);
-    0
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(data, data_len);
+    user_data.0.sha512_digest_update(buffer).into_code()
+}
+
+unsafe extern "C" fn sha512_digest_final_func(
+    _digest_context: *mut c_void,
+    output: *mut *mut signal_buffer,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    match user_data.0.sha512_digest_final() {
+        Ok(buf) => {
+            let buffer = Buffer::from(buf);
+            output.write(buffer.into_raw());
+            sys::SG_SUCCESS as c_int
+        },
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn sha512_digest_cleanup_func(
+    _digest_context: *mut c_void,
+    user_data: *mut c_void,
+) {
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    user_data.0.sha512_digest_cleanup();
+}
+
+unsafe extern "C" fn encrypt_func(
+    output: *mut *mut signal_buffer,
+    cipher: c_int,
+    key: *const u8,
+    key_len: usize,
+    iv: *const u8,
+    iv_len: usize,
+    plaintext: *const u8,
+    plaintext_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!user_data.is_null());
+    assert!(!key.is_null());
+    assert!(!iv.is_null());
+    assert!(!plaintext.is_null());
+
+    let key = slice::from_raw_parts(key, key_len);
+    let iv = slice::from_raw_parts(iv, iv_len);
+    let data = slice::from_raw_parts(plaintext, plaintext_len);
+    let signal_cipher_type = SignalCipherType::from(cipher);
```

Comment:
I feel like we should use `TryFrom` here and return an `InternalError::Unknown` or `InternalError::InvalidArgument`. That way we're explicitly dealing with the error case instead of panicking across the FFI boundary (which is UB, as far as I know).